src/preamble.js

```diff
@@ -419,21 +419,31 @@ function stringToAscii(str, outPtr) {
   return writeAsciiToMemory(str, outPtr, false);
 }
 
-// Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
+// Given a pointer 'ptr' to a UTF8-encoded string in the given array that contains uint8 values, returns
 // a copy of that string as a Javascript String object.
+// If len is passed, ptr points to an optionally null-terminated string of up to 'len' bytes.  If len is
+// not passed, ptr must point to a null-terminated string.
 
 #if TEXTDECODER
 var UTF8Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf8') : undefined;
 #endif
-function UTF8ArrayToString(u8Array, idx) {
+function UTF8ArrayToString(u8Array, idx, len) {
 #if TEXTDECODER
   var endPtr = idx;
+  // If len parameter is not passed, we need to find the length of the string.
   // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
   // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
-  while (u8Array[endPtr]) ++endPtr;
+  if (len == undefined) while (u8Array[endPtr]) ++endPtr;
+  else endPtr += len;
 
   if (endPtr - idx > 16 && u8Array.subarray && UTF8Decoder) {
+#ifdef USE_PTHREADS
```

Comment:
`#if USE_PTHREADS` instead of `#ifdef USE_PTHREADS`

Reply:
np, though I think the preprocessor doesn't care - it looks like it would accept any token starting with 'if' in fact :-)